### PR TITLE
qBittorrent's Homebrew formula.

### PR DESCRIPTION
**Pull request by ainstushar**
My 1st formula. For qBittorrent. Please let me know if I've screwed up.

Big thanks to @mistydemeo for spoon-feeding me these steps.

### CODE REVIEW

File: Library/Formula/qbittorrent.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+class Qbittorrent < Formula
+  desc "Qt based free and reliable P2P Bittorrent client"
+  homepage 'http://www.qbittorrent.org'
```
Comment (bfontaine) [inline Library/Formula/qbittorrent.rb]: Please use double-quotes everywhere.

File: Library/Formula/qbittorrent.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  url 'https://github.com/qbittorrent/qBittorrent/archive/release-3.2.3.tar.gz'
+  sha256 "816d611a0af50c469f52946c94b88c003f908ee031ee0e55a45273e4c64897c5"
+  head 'https://github.com/qbittorrent/qBittorrent.git',
+    :branch => ENV['QBT_BRANCH'] || "master"
```
Comment (bfontaine) [inline Library/Formula/qbittorrent.rb]: I don’t think we should support this.

File: Library/Formula/qbittorrent.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  def install
+    (buildpath/"src/gui/geoIP").install resource("geoIP") if build.stable?
+
+    system "./configure", "--disable-debug",
+                          "--disable-dependency-tracking",
+                          "--prefix=#{prefix}",
+                          "--disable-qt-dbus"
+    system "make"
+    system "macdeployqt src/qBittorrent.app"
```
Comment (bfontaine) [inline Library/Formula/qbittorrent.rb]: Please split this into one string per argument: `system "macdeployqt", "src/qBittorrent.app"`.

File: Library/Formula/qbittorrent.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    :branch => ENV['QBT_BRANCH'] || "master"
+
+  depends_on "pkg-config" => :build
+  depends_on "qt" => :build
+  depends_on "libtorrent-rasterbar" => :build
+  depends_on "boost" => :build
+  depends_on "gnu-sed" => :build
+
+  resource "geoIP" do
+    url "http://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz"
```
Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: This changes enormously frequently. It'll break on from-source builds. We'll need to find a less frequent mirror; Debian have one.

File: Library/Formula/qbittorrent.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                          "--disable-qt-dbus"
+    system "make"
+    system "macdeployqt src/qBittorrent.app"
+    prefix.install "src/qBittorrent.app"
+  end
+
+  test do
+    system "#{prefix}/qBittorrent.app/Contents/MacOS/qbittorrent", "-v"
+  end
+
```
Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: Don't need the empty line here.

File: Library/Formula/qbittorrent.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+class Qbittorrent < Formula
+  desc "Qt based free and reliable P2P Bittorrent client"
+  homepage 'http://www.qbittorrent.org'
+  url 'https://github.com/qbittorrent/qBittorrent/archive/release-3.2.3.tar.gz'
+  sha256 "816d611a0af50c469f52946c94b88c003f908ee031ee0e55a45273e4c64897c5"
+  head 'https://github.com/qbittorrent/qBittorrent.git',
+    :branch => ENV['QBT_BRANCH'] || "master"
+
+  depends_on "pkg-config" => :build
+  depends_on "qt" => :build
```
Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: This isn't likely to be a buildtime only dependency.

File: Library/Formula/qbittorrent.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  depends_on "libtorrent-rasterbar" => :build
+  depends_on "boost" => :build
```
Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: Don't think this is either, from the last time I build QBT from source.

Comment (mistydemeo) [inline Library/Formula/qbittorrent.rb]: `macdeployqt` embeds all of the libraries into the built app bundle.

Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: I had some problems with this before; I briefly supplied a few test `dmg` packages for upstream & the deploy process didn't quite work as it should have. Perhaps the root cause has been fixed upstream, or I just really sucked :smile:

File: Library/Formula/qbittorrent.rb (hunk 4 above)
Comment (ainstushar) [inline Library/Formula/qbittorrent.rb]: Yeah I discussed that with @mistydemeo and you are right. Maybe finding a different source would be helpful. Discussed not supplying sha256 for it. Any other suggestions might be helpful.

File: Library/Formula/qbittorrent.rb (hunk 2 above)
Comment (ainstushar) [inline Library/Formula/qbittorrent.rb]: @bfontaine, just curiosity, why is it not a good idea to support branch thingy?

File: Library/Formula/qbittorrent.rb (hunk 4 above)
Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: It's difficult, to be honest. Debian hack around this in their `geoip` package by doing this in the bootstrap:

```
mkdir -p $DIR/data
if [ ! -f $DIR/data/GeoIP.dat  ]; then
      curl http://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz | gzip -d > $DIR/data/GeoIP.dat
fi
```

Not ideal, really. My gut feeling would be for Homebrew to mirror semi-up-to-date GeoIP.dat "releases", but the attribution element of the license requirement makes that slightly tricky.

Comment (ainstushar) [inline Library/Formula/qbittorrent.rb]: They are updating their db only once a month. TBH I don't really have a problem with leaving this in there. I am sure qBittorrent devs can include attribution in their "About" section in app. That should cover it. Just to make sure, omitting sha256 is a big deal? (I am new at this so sorry in advance). I mean I could just have a script running on 1st Tuesday of each month updating hash...

Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: Kind of a big deal. We warn every time if a checksum isn't attached to a resource in a formula, unless that resource is from a git checkout or such.

```
==> Downloading http://geolite.maxmind.com/download/geoip/database/GeoLiteCountry/GeoIP.dat.gz
######################################################################## 100.0%
Warning: Cannot verify integrity of qbittorrent--geoIP-3.2.3.gz
A checksum was not provided for this resource
For your reference the SHA256 is: c03b4b8e10d67626fb3cd67cb338d186ce49273be60519dc4ad2ff073437bb87
```

Comment (ainstushar) [inline Library/Formula/qbittorrent.rb]: OK, so updating this formula on 1st Tuesday of each month with appropriate checksum is fine right?

Comment (DomT4) [inline Library/Formula/qbittorrent.rb]: I guess so, if nobody else objects.

Comment (dunn) [inline Library/Formula/qbittorrent.rb]: Still a little leery.  There'll be a short period once a month when installations fail.  And what if you forget?

Comment (ainstushar) [inline Library/Formula/qbittorrent.rb]: I was thinking about just running a cron job that monitors if the file changes and then automatically update the formula. But I see that if a user doesn't update formula, then they are screwed. I can't just keep a copy on github and distribute it due to their licensing. Other option is to keep an outdated copy in the code so we don't have to worry about adding it using formula.

Comment (bfontaine) [inline Library/Formula/qbittorrent.rb]: > But I see that if a user doesn't update formula, then they are screwed

They have a local copy that was downloaded by the formula so why would they be screwed?

Comment (sledgehammer999) [inline Library/Formula/qbittorrent.rb]: Guys, just so you know that url stays always the same. And that file updates only once a month.
Anyway, as discussed in our bugtracker from v3.3.0 you don't need to include geoip db.

Comment (mikemcquaid) [inline Library/Formula/qbittorrent.rb]: We need to work something out that doesn't rely on someone updating this once a month.

**Comment by mikemcquaid** [comment]
I think this would be a better fit for Homebrew Cask given that upstream provides a binary installer for QBittorrent.

**Comment by ainstushar** [comment]
@mikemcquaid There's already a cask formula available if user wants to install qBittorrent without compiling it. I made this formula so that people who wanted to test out the latest code on github can do so with ``--HEAD.`` As a matter of fact, I talked with some folks in homebrew cask irc and they recommended that I create a formula here. :) Was I told wrong?

**Comment by mikemcquaid** [comment]
I wouldn't say you're told wrong but I think I probably have a different opinion there. I'm OK with another maintainer merging this, though.

**Comment by mikemcquaid** [comment]
Maintainers: this may be a better fit for https://github.com/Homebrew/homebrew-gui if we decide to go ahead with that. @ainstushar Just ignore this for now.

**Comment by DomT4** [comment]
Going to pass on this *for now*, per:

> Anyway, as discussed in our bugtracker from v3.3.0 you don't need to include geoip db.

Once we reach that point and don't have to vendor in a file where the checksum is going to change and we have to require people to be very on-point in updating it every 30 days we'll review this again, if desired.

**Pull request closed without merging**